### PR TITLE
quic tpu: create fewer timeout futures

### PR DESCRIPTION
**Pull request by alessandrod**
Use a cancellation token instead of polling an `Arc<bool>` to exit connection tasks. Before this change we used to create a timeout future for each tx, and immediately cancel it since virtually all connections always have incoming txs pending. Cancelling timeout futures is expensive as it goes inside the tokio driver and takes a mutex. On high load (1M tps), cancelling timeout futures takes 8% (!) of run time.

With this change we create a cancellation token _once_ when a connection is established and that's it - no overhead after that.

### CODE REVIEW

File: streamer/src/nonblocking/quic.rs [hunk 1]
```diff
@@ -1046,107 +1046,114 @@ async fn handle_connection(
...
-                        let throttle_duration = STREAM_THROTTLING_INTERVAL
-                            .saturating_sub(throttle_interval_start.elapsed());
-
-                        if !throttle_duration.is_zero() {
-                            debug!("Throttling stream from {remote_addr:?}, peer type: {:?}, total stake: {}, \
+    loop {
+        // Wait for new streams. If the peer is disconnected we get a cancellation signal and stop
+        // the connection task.
+        let mut stream = select! {
+            stream = connection.accept_uni() => match stream {
```
Comment (alessandrod) [inline streamer/src/nonblocking/quic.rs:1053]: Here we don't need the timeout. Even if a staked peer opens a connection and never sends anything, they can open maximum 8 of them, and they'll get flushed out by the LRU logic.

File: streamer/src/nonblocking/quic.rs [hunk 2]
```diff
@@ -1046,107 +1046,114 @@ async fn handle_connection(
...
+        let stream_load_ema = stream_load_ema.clone();
+        tokio::spawn(async move {
+            let mut maybe_batch = None;
+            loop {
+                // Read the next chunk
+                let chunk = match tokio::select! {
+                    // Wait up to `wait_for_chunk_timeout`. If we don't get a chunk before then, we
+                    // assume the stream is dead and stop the stream task. This can only happen if
+                    // there's severe packet loss or the peer stop sending for whatever reason.
+                    chunk = tokio::time::timeout(
```
Comment (alessandrod) [inline streamer/src/nonblocking/quic.rs:1112]: Here we still need a timeout, because we want to limit the number of open-but-unfinished streams we let a peer open. I have a tokio patch coming to make this faster.

**Review by alessandrod — COMMENTED** [review]
I've highlighted the main changes in the PR. The rest is whitespace change because of some reduced nesting.

**Review by KirillLykov — DISMISSED** [review]
Looks good to me: it replaces `Arc<AtomicBool>` with `CancelationToken` and all operations in `select!` are cancel-safe.

File: programs/sbf/Cargo.lock [hunk 3]
```diff
@@ -6837,6 +6837,7 @@ dependencies = [
  "solana-transaction-metrics-tracker",
  "thiserror",
  "tokio",
+ "tokio-util 0.7.1",
```
Comment (lijunwangs) [inline programs/sbf/Cargo.lock:6891]: Any reason why the versions are different? Can we use 0.7.12 for both?

File: streamer/src/nonblocking/quic.rs [hunk 4]
```diff
@@ -65,10 +65,10 @@ use {
         task::JoinHandle,
         time::{sleep, timeout},
     },
+    tokio_util::sync::CancellationToken,
 };
 
-const WAIT_FOR_STREAM_TIMEOUT: Duration = Duration::from_millis(100);
-pub const DEFAULT_WAIT_FOR_CHUNK_TIMEOUT: Duration = Duration::from_secs(10);
+pub const DEFAULT_WAIT_FOR_CHUNK_TIMEOUT: Duration = Duration::from_secs(1);
```
Comment (lijunwangs) [inline streamer/src/nonblocking/quic.rs]: What is the reasoning changing this timeout value from 10s to 1s?

Comment (alessandrod) [reply streamer/src/nonblocking/quic.rs]: old timeout makes no sense - a validator has 1.6s worth of leader slots, what are we waiting for 10 seconds?

Comment (lijunwangs) [reply streamer/src/nonblocking/quic.rs]: That was conservative -- it does forward if passing the leader slot. Note 1 < 1.6s. I would like to keep the old timeout to reduce the net change impact. If we are changing it, maybe a separate PR?

Comment (alessandrod) [reply streamer/src/nonblocking/quic.rs]: > Note 1 < 1.6s.

Yes, this is intentional. Waiting _ten seconds_ for a single _chunk_ of _one_ transaction makes no sense. The only way it can happen is if the peer is being intentionally malicious and has opened one (or more) streams and is not sending data in them. 

>  it does forward if passing the leader slot

I'll check this but hopefully we don't do it for that long because that would also make no sense

Comment (alessandrod) [reply streamer/src/nonblocking/quic.rs]: > I would like to keep the old timeout to reduce the net change impact. If we are changing it, maybe a separate PR?

To do this I'd have to make it do the old thing of sleeping in WAIT_FOR_CHUNK_TIMEOUT / 10 intervals or the tests take forever. So I'd have to add back the old code complexity and create * 10 the number of timeout futures - in a PR trying to reduce the number of timeout futures. To me "this is what we used to do" is not a good enough argument to justify the complexity and worse perf.

Comment (t-nelson) [reply streamer/src/nonblocking/quic.rs]: what if you reduce the timeout in a predecessor pr?

Comment (t-nelson) [reply streamer/src/nonblocking/quic.rs]: or just separate commit and we merge w/o squashing

Comment (alessandrod) [reply streamer/src/nonblocking/quic.rs]: wait, is not squashing even an option? 🤯

Comment (t-nelson) [reply streamer/src/nonblocking/quic.rs]: i know a guy who can make it one, temporarily

File: programs/sbf/Cargo.lock (hunk 3 above)
Comment (alessandrod) [reply programs/sbf/Cargo.lock:6891]: Because this is how dependency resolution works. Something in the dep graph requires <= 0.7.1 and because 0.7.12 and 0.7.1 are semver compatible it's using that. I didn't go and edit Cargo.lock - cargo did this change. If you look at Cargo.lock, it's already using 0.7.1 in many places.

File: streamer/src/nonblocking/quic.rs (hunk 4 above)
Comment (alessandrod) [reply streamer/src/nonblocking/quic.rs]: done, split in two commits, do your magic now trent

File: programs/sbf/Cargo.lock (hunk 3 above)
Comment (lijunwangs) [reply programs/sbf/Cargo.lock:6891]: Strange both are about solana-streamer's dependencies.

File: streamer/src/nonblocking/quic.rs (hunk 4 above)
Comment (lijunwangs) [reply streamer/src/nonblocking/quic.rs]: I am questioning the 1s is too small and a lot of more txns may timeout due to this. At least it need to be within the leader slot. We do need to be careful fixing one issue and introducing other potential issues.

Comment (alessandrod) [reply streamer/src/nonblocking/quic.rs]: Could you walk me through what would timeout and for what reason?

Comment (lijunwangs) [reply streamer/src/nonblocking/quic.rs]: Screenshot 2024-10-15 at 10 52 34 AM
see this in the mainnet with the current 10s timeout value. The stream can timeout due to network conditions (client, in between, server).  I also wonder why the cancel mechanism you implemented here cannot be used.

Comment (alessandrod) [reply streamer/src/nonblocking/quic.rs]: > see this in the mainnet with the current 10s timeout value. The stream can timeout due to network conditions (client, in between, server).

We want to timeout garbage connections as soon as possible, we're trying to build a chain that goes fast. If you're sending txs from your dialup at home you should get disconnected.

Also this is a timeout for _one single stream_, a connection can still send other streams. So in case of legitimate packet loss (one or a few transactions), we wait up to 1s, otherwise we move on and keep reading other transactions from the same peer, who will receive a reset stream error and presumably retry the failed transactions.

What you're seeing in those metrics is probably peers timing out because the whole connection times out, and then because we're currently extremely bad and read many streams in parallel, we increment the metric many times.

So tldr: on legit packet loss, 1s seems more than enough (assuming decent RTT <=200ms). When a peer falls off the planet and the connection just stops sending any data, timing out as soon as possible is the best thing we can do to allocate resources to peers who are actually sending transactions.

> I also wonder why the cancel mechanism you implemented here cannot be used. 

I will let you role play as hacker for one day to answer that question!

Comment (lijunwangs) [reply streamer/src/nonblocking/quic.rs]: I am not comfortable with the big change of the timeout value. Can we set it to 2s for now as an interim approach and observe in the production of its impacts?

**Review by lijunwangs — DISMISSED** [review]

**Comment by KirillLykov** [comment]
In CI `test_quic_server_multiple_connections_on_single_client_endpoint` failed with message:
```
thread 'nonblocking::quic::test::test_quic_server_multiple_connections_on_single_client_endpoint' panicked at streamer/src/nonblocking/quic.rs:1880:9:
--
  | assertion `left == right` failed
  | left: 0
  | right: 2
```
We check  that the number of closed connections is 2 while it is 0 here

**Comment by alessandrod** [comment]
> In CI `test_quic_server_multiple_connections_on_single_client_endpoint` failed with message:
> 
> ```
> thread 'nonblocking::quic::test::test_quic_server_multiple_connections_on_single_client_endpoint' panicked at streamer/src/nonblocking/quic.rs:1880:9:
> --
>   | assertion `left == right` failed
>   | left: 0
>   | right: 2
> ```
> 
> We check that the number of closed connections is 2 while it is 0 here

From the code I removed

```
           // The min is to guard against a value too small which can wake up unnecessarily
            // frequently and wasting CPU cycles. The max guard against waiting for too long
            // which delay exit and cause some test failures when the timeout value is large.
            // Within this value, the heuristic is to wake up 10 times to check for exit
            // for the set timeout if there are no data.
```

This test is racy and was relying on sleep time being a lot larger than timeout time... I'll fix it properly tomorrow.

File: streamer/src/nonblocking/quic.rs [hunk 5]
```diff
@@ -1815,7 +1811,7 @@ pub mod test {
         join_handle.await.unwrap();
     }
 
-    #[tokio::test]
+    #[tokio::test(flavor = "multi_thread")]
```
Comment (alessandrod) [inline streamer/src/nonblocking/quic.rs:1814]: By default tokio::test uses the single threaded runtime. This is why the huge timeout was needed, because if things were scheduled so that the thread went inside epoll_wait(), then it'd get stuck in there for however long the timeout is, messing with the (fragile) timing assumptions made by the test.

File: streamer/src/nonblocking/quic.rs [hunk 6]
```diff
@@ -1864,11 +1860,13 @@ pub mod test {
             CONNECTION_CLOSE_CODE_DROPPED_ENTRY.into(),
             CONNECTION_CLOSE_REASON_DROPPED_ENTRY,
         );
-        // Wait long enough for the stream to timeout in receiving chunks
-        let sleep_time = Duration::from_secs(1).min(WAIT_FOR_STREAM_TIMEOUT * 1000);
-        sleep(sleep_time).await;
 
-        assert_eq!(stats.connection_removed.load(Ordering::Relaxed), 1);
+        let start = Instant::now();
+        while stats.connection_removed.load(Ordering::Relaxed) != 1 {
```
Comment (alessandrod) [inline streamer/src/nonblocking/quic.rs:1865]: Here and below, I made it loop waiting for the expected condition. If the expected conditions are never reached, the test fails by timing out. I still put an upper bound (1s) on how long it's ok to wait for the server to close the connections. 1s should be enough to account for terrible scheduling perf in CI if things get overloaded.

**Review by lijunwangs — APPROVED** [review]
LGTM
